add new page about TLS (#9412)
pantheon-systems/documentation: source/content/tls-compatibility.md

A web application uses TLS when communicating with another server or service, such as an API or a payment gateway. As such, it is essential to ensure that your site is compatible with the latest TLS standards to protect your users' data and maintain the security of your site.

A TLS connection is initiated by a "handshake". You can think of this as two computers introducing themselves to each other. As part of the TLS handshake, the two servers will acknowledge each other, determine what cryptographic algorithms to use, exchange session keys and handle authentication. This is when the version of TLS encryption is determined.

## TLS Compatibility

TLS 1.0 and 1.1 were published in 1999 and 2006, respectively. Both were deprecated in 2021 and are considered outdated and insecure. Pantheon servers support TLS 1.2 and 1.3 connections. This means any outgoing HTTP requests being made from the _application level_ must be with servers that also support TLS 1.2 or 1.3.

> Pantheon application servers will _reject_ connections to applications that do not support TLS 1.2+. <!--TODO: Update to note *when* we will no longer allow TLS 1.2+ connections? -->

Author: @scottbuscemi If we're publishing this doc before the change, should we put an (anticipated) date?

Reviewer: We don't have a firm date yet for the switchover. In the meantime, we can convey that this will be enforced in late 2025.

Drupal users can use the [Pantheon Drupal TLS Compatibility Checker](https://github.com/pantheon-systems/drupal_tls_checker) to scan their site for outgoing TLS connections. This module supports Drupal 10/11 and can be run from the command line using Drush or in the admin at `/admin/config/development/tls-checker`. If any URLs are found in the `modules` or `themes` directories that do not support TLS 1.2 or 1.3 connections, a list is provided on that page (or shown in the terminal).

View the project [README](https://github.com/pantheon-systems/drupal_tls_checker/blob/main/README.md) for more detailed information about customizing the checks, installation and usage.

</Tab>

<Tab title="Command Line" id="check-compatibility-cli">

Individual URLs can be checked for TLS 1.2/1.3 compatibility with the `curl` command.

```bash{promptUser: user}
curl --tlsv1.2 --head --connect-timeout 10 "<your url to check>"
```

If you receive a 200 status code in the response (e.g. `HTTP/2 200`) along with header information, the check succeeded. If you receive an error (e.g. `TLS connect error`), it means the server you are attempting to make a connection to does not support that version of TLS.

> A more robust example that scans all the files in a particular path can be found [in this Gist](https://gist.github.com/scottbuscemi/8915d6d53f34f5012406ef5d21b222a1).

Author: @scottbuscemi okay to link to your gist?

Reviewer: I don't think that's a sustainable choice for the docs. Better to just copy it into the docs.

reviewed: "2025-03-07"
---

This document describes TLS encryption, when it's used, and how to test TLS compatibility.

## What is TLS?

Transport Layer Security (TLS) is a widely adopted cryptographic security protocol that provides secure communication over a computer network. It is the successor to Secure Sockets Layer (SSL) and is used to secure connections between web servers and clients. One of the primary uses of TLS is to encrypt the communication between web applications and servers, such as when a user logs into a website or submits a form.

A web application uses TLS when communicating with another server or service, such as an API or a payment gateway. As such, it is essential to ensure that your site is compatible with the latest TLS standards to protect your users' data and maintain the security of your site.

A TLS connection is initiated by a "handshake". You can think of this as two computers introducing themselves to each other. As part of the TLS handshake, the two servers will acknowledge each other, determine what cryptographic algorithms to use, exchange session keys and handle authentication. This is when the version of TLS encryption is determined.

> ## TLS Compatibility

Reviewer: ```suggestion
## Pantheon Platform TLS Compatibility
```

TLS 1.0 and 1.1 were published in 1999 and 2006, respectively. Both were deprecated in 2021 and are considered outdated and insecure. Pantheon servers support TLS 1.2 and 1.3 connections. This means any outgoing HTTP requests being made from the _application level_ must be with servers that also support TLS 1.2 or 1.3.

> Pantheon application servers will _reject_ connections to applications that do not support TLS 1.2+. <!--TODO: Update to note *when* we will no longer allow TLS 1.2+ connections? -->

Author: ```suggestion
Pantheon application servers will _reject_ connections to applications that do not support TLS 1.2+ beginning in late 2025.
```

<Alert title="What's affected?" type="info">

**Application requests** include things like cURL or remote HTTP requests (e.g. `wp_remote_get` in WordPress) against a server or service. They do _not_ include requests for remotely hosted JavaScript or CSS files that are loaded when a webpage is rendered. (Your web browser handles the TLS handshake for these types of requests.)

</Alert>

### How do I check TLS compatibility for my site?

You can check your whether site's codebase makes HTTP connections to servers that support TLS 1.2 or 1.3 by using the WordPress TLS Compatibility Checker plugin or via the command line.

<TabList>

<Tab title="WordPress" id="check-compatibility-wp" active={true}>

> WordPress users can use the [Pantheon WP TLS Compatibility Checker](https://github.com/pantheon-systems/wp-tls-compatibility-checker) to scan their site for outgoing TLS connections. The plugin can be run from the command line using WP-CLI or in the admin via a new TLS Compatibility page under the Tools menu. If any URLs are found that do not support TLS 1.2 or 1.3 connections, a list is provided on the TLS Compatibility page (or shown in the terminal).

Reviewer: ```suggestion
WordPress users can use the [Pantheon WP TLS Compatibility Checker](https://github.com/pantheon-systems/wp-tls-compatibility-checker) plugin to scan their site for outgoing TLS connections. The plugin can be run from the command line using WP-CLI or in the admin via a new TLS Compatibility page under the Tools menu. If any URLs are found that do not support TLS 1.2 or 1.3 connections, a list is provided on the TLS Compatibility page (or shown in the terminal).
```
